backend/src/auth/jwt/jwt.service.ts

```diff
@@ -0,0 +1,106 @@
+import { ForbiddenException, Injectable } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { JwtService } from '@nestjs/jwt';
+import { CookieOptions, Response } from 'express';
+import { UsersService } from 'src/users/users.service';
+import { User } from '@prisma/client';
+import { AuthTokens, JwtPayload } from '../dto/jwt.dto';
+import ms from 'ms';
+import * as argon2 from 'argon2';
+
+@Injectable()
+export class JwtAuthService {
+  constructor(
+    private jwtService: JwtService,
+    private usersService: UsersService,
+    private configService: ConfigService,
+  ) {}
+
+  private accessTokenExpiration: string = this.configService.get(
+    'ACCESS_TOKEN_EXPIRATION',
+    '1h',
+  );
+
+  private refreshTokenExpiration: string = this.configService.get(
+    'REFRESH_TOKEN_EXPIRATION',
+    '1d',
+  );
+
+  private generatePayload(user: User) {
+    return {
+      sub: user.login,
+      email: user.email,
+    } as JwtPayload;
+  }
+
+  private async updateRefreshToken(user: User, refreshToken: string) {
+    const hash = await argon2.hash(refreshToken, {
+      secret: Buffer.from(this.configService.get('HASH_PEPPER')),
+    });
+
+    await this.usersService.update(user.login, {
+      refreshToken: hash,
+    });
+  }
+
+  async generateJwt(user: User) {
+    const payload = this.generatePayload(user);
+    const accessToken = this.jwtService.sign(payload, {
+      expiresIn: this.accessTokenExpiration,
+      secret: this.configService.get('JWT_SECRET'),
+    });
+    const refreshToken = this.jwtService.sign(payload, {
+      expiresIn: this.refreshTokenExpiration,
+      secret: this.configService.get('JWT_SECRET'),
+    });
+
+    await this.updateRefreshToken(user, refreshToken);
+
+    return {
+      accessToken,
+      refreshToken,
+    };
+  }
+
+  private generateCookieOptions(): CookieOptions {
+    // TODO retrieve domain from config later on
+    const domain =
+      this.configService.get('NODE_ENV') !== 'production' ? 'localhost' : '';
```

Comment:
This is a form of differentiating between environments but this is not a definite implementation, since the deployment will require an actual domain, which we can retrieve from another env var later on.

Reply:
ACK! Good take